apache2/waf_logging/waf_log_util_internal.h

```diff
@@ -1,9 +1,100 @@
 #ifndef _WAF_LOG_UTIL_INTERNAL_HEADER
 #define _WAF_LOG_UTIL_INTERNAL_HEADER
 
-#include <google/protobuf/util/json_util.h>
-#include"waf_format.pb.h"
 #include <fstream>
+#include <unordered_map>
 #include <exception>
+#include "../../jsoncons/include/jsoncons/json.hpp"
+#include "../../jsoncons/include/jsoncons/json_options.hpp"
+
+// For convenience
+using jsoncons::ojson;
+using jsoncons::json_options;
+using jsoncons::line_split_kind;
+
+namespace ns {
+
+    struct detail
+    {
+        std::string message;
+        std::string data;
+        std::string file;
+        std::string line;
+    };
+
+    struct property
+    {
+        std::string instanceId;
+        std::string clientIp;
+        std::string clientPort;
+        std::string requestUri;
+        std::string ruleSetType;
+        std::string ruleSetVersion;
+        std::string ruleId;
+        std::string message;
+        std::string action;
+        std::string site;
+        detail details;
+        std::string hostname;
+        std::string transactionId;
+    };
+
+    class waf_log_object
+    {
+        std::string resourceId;
+        std::string operationName;
+        std::string category;
+        property properties;
+
+        // Make json_type_traits specializations friends to give accesses to private members
+        JSONCONS_TYPE_TRAITS_FRIEND;
+    public:
+        waf_log_object(){
+        }
+
+        waf_log_object(const std::string& resourceId, const std::string& operationName, const std::string& category, const property& properties)
+            : resourceId(resourceId), operationName(operationName), category(category), properties(properties) {                
+        }
+        
+        std::string to_json_string(){
+            ojson detail;
+            ojson property;
+            ojson waf_log;
+
+            json_options options;
+            options.indent(0);
+            options.new_line_chars(" ");
+            std::stringstream buffer;
+            
+            property["instanceId"] = properties.instanceId;
```

Comment:
This sequence of explicit assignments looks saddening. I would expect `jsoncons` provides us with a way to register hooks for our fields in the structure that automatically accesses them with some common serializer - is it not the case?

With these we are not far from just manually contatenating strings into JSON and I think we want to avoid it.

Reply:
I think we should do something similar to this example: https://danielaparker.github.io/jsoncons/#A4

Reply:
it doesn't have corresponding function of ojson, which will keep the insertion order, that's why i manually concatenate the json

Reply:
I have modified the example from `jsoncons` to work with `ojson`. You can use it as a reference. It outputs JSON field in the same order that is used in JSONCONS_TYPE_TRAITS_DECL. If you re-arrange the names in it, the output changes accordingly. Give it a try.
```c++
#include <cassert>
#include <iostream>
#include <jsoncons/json.hpp>

namespace jc = jsoncons;

namespace ns {

    struct reputon
    {
        std::string rater;
        std::string assertion;
        std::string rated;
        double rating;

        friend bool operator==(const reputon& lhs, const reputon& rhs)
        {
            return lhs.rater == rhs.rater &&
                lhs.assertion == rhs.assertion &&
                lhs.rated == rhs.rated &&
                lhs.rating == rhs.rating;
        }

        friend bool operator!=(const reputon& lhs, const reputon& rhs)
        {
            return !(lhs == rhs);
        };
    };

    class reputation_object
    {
        std::string application;
        std::vector<reputon> reputons;

        // Make json_type_traits specializations friends to give accesses to private members
        JSONCONS_TYPE_TRAITS_FRIEND;
    public:
        reputation_object()
        {
        }
        reputation_object(const std::string& application, const std::vector<reputon>& reputons)
            : application(application), reputons(reputons)
        {}

        friend bool operator==(const reputation_object& lhs, const reputation_object& rhs)
        {
            if (lhs.application != rhs.application)
            {
                return false;
            }
            if (lhs.reputons.size() != rhs.reputons.size())
            {
                return false;
            }
            for (size_t i = 0; i < lhs.reputons.size(); ++i)
            {
                if (lhs.reputons[i] != rhs.reputons[i])
                {
                    return false;
                }
            }
            return true;
        }

        friend bool operator!=(const reputation_object& lhs, const reputation_object& rhs)
        {
            return !(lhs == rhs);
        };
    };

} // namespace ns

// Declare the traits. Specify which data members need to be serialized.
JSONCONS_TYPE_TRAITS_DECL(ns::reputon, rater, assertion, rated, rating);
JSONCONS_TYPE_TRAITS_DECL(ns::reputation_object, application, reputons);

int main()
{
    ns::reputation_object val("hiking", { ns::reputon{"HikingAsylum.example.com","strong-hiker","Marilyn C",0.90} });

    jc::json_options options;
    options.indent(0);
    options.new_line_chars(" ");

    std::string s;
    jc::encode_json(jc::ojson{}, val, s, options, jc::indenting::indent);
    std::cout << s << "\n";

    auto val2 = jc::decode_json<ns::reputation_object>(s);

    assert(val2 == val);
}

```